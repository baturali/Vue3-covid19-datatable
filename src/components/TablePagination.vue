<script setup lang="ts">
import { computed } from 'vue'
import type { TableOptions } from '../services/types'

export interface PaginationProps {
  tableOptions: TableOptions
  totalCount: number
  inputValue?: string
}
const paginationProps = defineProps<PaginationProps>()

const emitEvent = defineEmits<{
  (e: 'change-page', sign: boolean): void
}>()

const rowCount = computed(() => {
  return paginationProps.tableOptions.rowCount
})
const totalCount = computed(() => {
  return paginationProps.totalCount
})
const page = computed(() => {
  return paginationProps.tableOptions.page
})
const firstIndexOfPage = computed(() => {
  return page.value * rowCount.value
})
const inputValue = computed(() => {
  return paginationProps.inputValue
})
</script>

<template>
  <footer>
    <div class="filter" v-if="inputValue">
      Showing filtered results for <span class="input-value">"{{ inputValue }}"</span>
    </div>
    <div class="pagination" v-else>
      <div>
        Showing
        {{ firstIndexOfPage }} - 
        {{ totalCount < firstIndexOfPage + rowCount ? totalCount : firstIndexOfPage + rowCount}}
        of {{ totalCount < rowCount ? rowCount : totalCount }}
      </div>
      <div class="actions">
        <button
          :class="{ 'disabled': !page }"
          @click="page && emitEvent('change-page', false)"
        >
          Previous
        </button>
        <span class="page-indicator">Page: {{ page + 1}}</span>
        <button
          :class="{ 'disabled': !(((page + 1) * rowCount) < totalCount) }"
          @click="(((page + 1) * rowCount) < totalCount) && emitEvent('change-page', true)"
        >
          Next
        </button>
      </div>
    </div>
  </footer>
</template>

<style lang="postcss" scoped>
footer {
  @apply w-full pt-4;
}

.filter {
  @apply text-white text-lg font-bold text-center w-full;

  & .input-value {
    @apply text-purple-700;
  }
}

.pagination {
  @apply flex flex-row w-full items-center justify-between;

  & .actions {
    @apply flex items-center gap-x-6;

    & .page-indicator {
      @apply text-lg text-white font-bold;
    }
  }

  button {
    @apply bg-purple-500 rounded-xl py-2 px-3 transition-all text-white;

    &:hover {
      @apply bg-purple-700;
    }
  }
  button.disabled {
    @apply bg-gray-400 cursor-not-allowed;
  }
}
</style>