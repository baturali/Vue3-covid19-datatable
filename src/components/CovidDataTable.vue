<script setup lang="ts">
import { ref, reactive, computed, watch } from 'vue'
import type { SummaryDTO, SummaryCountriesDTO, TableOptions } from '../services/types'
import TablePagination from '@/components/TablePagination.vue'

export interface DataTableProps {
  summary?: SummaryDTO
  input?: string
}
const dataTableProps = defineProps<DataTableProps>()

const countries = ref<SummaryCountriesDTO[]>()
const inputValue = ref<string>()

watch(dataTableProps, (newVal, oldVal) => {
  if (newVal) { 
    countries.value = dataTableProps.summary?.Countries
    inputValue.value = dataTableProps.input
  }
}, { immediate: true })

const pageDataHandler = computed(() => {
  if (!countries.value) return []
  const { input } = dataTableProps
  const { page, rowCount } = tableOptions
  return countries.value
    .filter((country: SummaryCountriesDTO) => input ? country.Country.toLowerCase().includes(input.toLowerCase()) : country)
    .slice(input ? 0 : page * rowCount, input ? countries.value.length : (page + 1) * rowCount)
})

const tableOptions = reactive<TableOptions>({
  page: 0,
  rowCount: 20,
  headers: ['Country', 'TotalConfirmed', 'TotalDeaths', 'TotalRecovered']
})


function translatedLabel(label: string) {
  return label.replace(/([A-Z][a-z])/g,' $1').replace(/(\d)/g,' $1');
}

function getColumnSum(key: 'TotalConfirmed' | 'TotalDeaths' | 'TotalRecovered') {
  return pageDataHandler.value
    .map((row: SummaryCountriesDTO) => {
      return row[key]
    })
    .reduce((a: number, b: number) => a + b, 0)
}

function formatNumbers(value: number) {
  return value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",")
}

function changePage(sign: boolean) {
  sign ? tableOptions.page++ : tableOptions.page--
}
</script>

<template>
  <table> 
    <thead>
      <tr>
        <td v-for="header in tableOptions.headers" :key="header" :data-test="header">
          {{ translatedLabel(header) }}
        </td>
      </tr>
    </thead>
    <tbody>
      <tr v-for="(row, index) in pageDataHandler" :key="row.Slug" data-test="data-row">
        <td>{{ row.Country }}</td>
        <td :data-test="index === 0 ? 'first-confirmed' : 'confirmed'">
          {{ formatNumbers(row.TotalConfirmed) }}
        </td>
        <td>{{ formatNumbers(row.TotalDeaths) }}</td>
        <td>{{ formatNumbers(row.TotalRecovered) }}</td>
      </tr>
    </tbody>
    <tfoot>
      <tr>
        <td class="text-right"> Total: </td>
        <td data-test="total-confirmed">{{ formatNumbers(getColumnSum('TotalConfirmed')) }}</td>
        <td data-test="total-deaths">{{ formatNumbers(getColumnSum('TotalDeaths')) }}</td>
        <td data-test="total-recovered">{{ formatNumbers(getColumnSum('TotalRecovered')) }}</td>
      </tr>
    </tfoot>
  </table>
  <TablePagination
    :table-options="tableOptions"
    :total-count="countries ? countries.length : 0"
    :input-value="inputValue"
    @change-page="changePage"
  />
</template>

<style lang="postcss" scoped>
table {
  @apply w-full bg-white rounded-sm overflow-hidden;

  & tr > td {
    @apply px-4 py-2 w-1/4 text-black;
  }

  & thead > tr > td {
    @apply text-black text-xl border-b border-b-gray-400 pt-6 font-extrabold;
  }

  & tbody > tr:nth-child(odd) > td {
    @apply bg-gray-200;
  }

  & tfoot > tr > td {
    @apply bg-gray-700 font-extrabold text-white;
  }
}
</style>