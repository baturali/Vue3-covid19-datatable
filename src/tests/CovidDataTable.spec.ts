import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'

import CovidDataTable from '@/components/CovidDataTable.vue'
import type { SummaryDTO } from '../services/types'
import TablePagination from '@/components/TablePagination.vue'

const defaultMountOptions: { props: {
  summary?: SummaryDTO
  input?: string
} } = {
  props: {
    summary: {
      Countries: [
        {
          Country: 'Test country',
          CountryCode: 'Some code',
          Date: 'Today',
          ID: 'iftbbg3241vtvb',
          NewConfirmed: 64,
          NewDeaths: 0,
          NewRecovered: 0,
          Slug: 'test-country',
          TotalConfirmed: 201276,
          TotalDeaths: 7311,
          TotalRecovered: 1
        },
        {
          Country: 'A real country',
          CountryCode: 'Some code',
          Date: 'Today',
          ID: 'iftbbg3241vtvb',
          NewConfirmed: 64,
          NewDeaths: 0,
          NewRecovered: 0,
          Slug: 'a-real-country',
          TotalConfirmed: 1322,
          TotalDeaths: 9,
          TotalRecovered: 1313
        }
      ],
      Global: {
        Date: "2022-10-19T05:34:04.363Z",
        NewConfirmed: 533718,
        NewDeaths: 1706,
        NewRecovered: 0,
        TotalConfirmed: 623745794,
        TotalDeaths: 6563578,
        TotalRecovered: 0
      },
      ID: 'test-id',
      Message: ''
    },
    input: ''
  }
}

describe('CovidDataTable', () => {
  it('Renders properly', () => {
    const wrapper = mount(CovidDataTable)
    expect(wrapper.find('table')).toMatchSnapshot()
  })

  it('Should have the translated header values', () => {
    const wrapper = mount(CovidDataTable)
    const headers = [
      ['TotalConfirmed', 'Total Confirmed'],
      ['TotalDeaths', 'Total Deaths'],
      ['TotalRecovered', 'Total Recovered']
    ]
    headers.forEach(([header, translated]) => {
      expect(wrapper.find(`[data-test=${header}]`).text()).toBe(translated)
    })
  })

  it('Shows prepared rows with given props', () => {
    const wrapper = mount(CovidDataTable, defaultMountOptions)
    expect(wrapper.find('[data-test=data-row]')).toBeTruthy()
    expect(wrapper.findAll('[data-test=data-row]').length).toBe(2)
  })

  it('Displays accumulated "total" values', () => {
    const wrapper = mount(CovidDataTable, defaultMountOptions)
    const totalColumns = ['total-confirmed', 'total-deaths', 'total-recovered']
    totalColumns.forEach((column) => {
      expect(wrapper.find(`[data-test=${column}]`).text()).toBeTruthy()
    })
  })

  it('Filters the table list and returns matched row via user input', async () => {
    const wrapper = mount(CovidDataTable, defaultMountOptions)
    wrapper.setProps({
      input: 'Test '
    })
    await wrapper.vm.$nextTick()
    expect(wrapper.findAll('[data-test=data-row]').length).toBe(1)
  })

  it('Should change the page via Previous and Next buttons', async () => {
    const wrapper = mount(CovidDataTable, defaultMountOptions)
    const totalConfirmed = wrapper.find('[data-test=total-confirmed]').text()
    const children = wrapper.findComponent(TablePagination)
    children.vm.$emit('change-page', true)
    await wrapper.vm.$nextTick()
    const updatedTotalConfirmed = wrapper.find('[data-test=total-confirmed]').text()
    expect(totalConfirmed === updatedTotalConfirmed).toBe(false)
  })

  it('Should format numbers', () => {
    const wrapper = mount(CovidDataTable, defaultMountOptions)
    const totalConfirmed = wrapper.find('[data-test=first-confirmed]').text()
    expect(totalConfirmed !== defaultMountOptions.props.summary?.Countries[0].TotalDeaths.toString()).toBe(true)
  })
})
