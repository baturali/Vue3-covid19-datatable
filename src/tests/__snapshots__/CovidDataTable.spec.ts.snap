// Vitest Snapshot v1

exports[`CovidDataTable > Renders properly 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <table
    data-v-8a977408=""
  >
    <thead
      data-v-8a977408=""
    >
      <tr
        data-v-8a977408=""
      >
        
        <td
          data-test="Country"
          data-v-8a977408=""
        >
           Country
        </td>
        <td
          data-test="TotalConfirmed"
          data-v-8a977408=""
        >
           Total Confirmed
        </td>
        <td
          data-test="TotalDeaths"
          data-v-8a977408=""
        >
           Total Deaths
        </td>
        <td
          data-test="TotalRecovered"
          data-v-8a977408=""
        >
           Total Recovered
        </td>
        
      </tr>
    </thead>
    <tbody
      data-v-8a977408=""
    >
      
      
    </tbody>
    <tfoot
      data-v-8a977408=""
    >
      <tr
        data-v-8a977408=""
      >
        <td
          class="text-right"
          data-v-8a977408=""
        >
           Total: 
        </td>
        <td
          data-test="total-confirmed"
          data-v-8a977408=""
        >
          0
        </td>
        <td
          data-test="total-deaths"
          data-v-8a977408=""
        >
          0
        </td>
        <td
          data-test="total-recovered"
          data-v-8a977408=""
        >
          0
        </td>
      </tr>
    </tfoot>
  </table>,
}
`;
